app/src/main/java/com/tatar/currencify/ui/feature/CurrencyConversionActivity.kt

```diff
@@ -0,0 +1,234 @@
+        observeState(binding)
+        observeEvents(binding)
+    }
+
+    private fun provideDependencies() {
+        DaggerCurrencyConversionComponent.factory()
+            .create(CurrencifyApp.appComponent(this))
+            .inject(this)
+    }
+
+    private fun setViewModel() {
+        currencyConversionViewModel =
+            ViewModelProvider(this, viewModelFactory).get(CurrencyConversionViewModel::class.java)
+    }
+
+    private fun setupAnimation(binding: ActivityCurrencyConversionBinding) {
+        blinkAnim = AnimationUtils.loadAnimation(this, R.anim.blink)
+
+        blinkAnim.setAnimationListener(object : Animation.AnimationListener {
+            override fun onAnimationStart(animation: Animation) {
+                binding.currencyResponderDivider.setBackgroundColor(
+                    ContextCompat.getColor(
+                        this@CurrencyConversionActivity,
+                        R.color.colorAccent
+                    )
+                )
+            }
+
+            override fun onAnimationEnd(animation: Animation) {
+                binding.currencyResponderDivider.setBackgroundColor(
+                    ContextCompat.getColor(
+                        this@CurrencyConversionActivity,
+                        R.color.colorTextLight
+                    )
+                )
+            }
+
+            override fun onAnimationRepeat(animation: Animation) {}
+        })
+    }
+
+    private fun setupInteractions(binding: ActivityCurrencyConversionBinding) {
+        binding.tryAgainBtn.setOnClickListener {
+            currencyConversionViewModel.onTryAgainClick()
+        }
+
+        binding.responderCurrencyContainer.setOnClickListener {
+            currencyConversionViewModel.onResponderContainerClick()
+        }
+
+        binding.responderCurrencyAmountEt.addTextChangedListener(object : TextWatcher {
+            override fun afterTextChanged(s: Editable) {}
+            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
+            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
+                currencyConversionViewModel.onAmountChange(s.toString())
+            }
+        })
+    }
+
+    private fun setupRecyclerview(binding: ActivityCurrencyConversionBinding) {
+        with(binding) {
+            convertedCurrenciesAdapter.setItemClickListener(this@CurrencyConversionActivity)
+            currencyConversionRv.adapter = convertedCurrenciesAdapter
+
+            currencyConversionRv.layoutManager = LinearLayoutManager(
+                this@CurrencyConversionActivity,
+                LinearLayoutManager.VERTICAL,
+                false
+            )
+        }
+    }
+
+    private fun observeState(binding: ActivityCurrencyConversionBinding) {
+        currencyConversionViewModel.loadingState.observe(
+            this,
+            Observer { isLoading ->
+                if (isLoading) {
```

Comment:
I'd prefer `toggle` methods instead of `hide` and `show` methods. This way you could have less method and your Observer function would be shorter. I know that's personal preference but I would make my helper functions more crowded than my observer,

Reply:
And to show how I would do it 
```
    private fun toggleLoadingIndicator(show: boolean, binding: ActivityCurrencyConversionBinding) {
        binding.progressBar.visibility = show ? View.VISIBLE : View.GONE
    }
```
instead of 
```
    private fun displayLoadingIndicator(binding: ActivityCurrencyConversionBinding) {
        binding.progressBar.visibility = View.VISIBLE
    }

    private fun hideLoadingIndicator(binding: ActivityCurrencyConversionBinding) {
        binding.progressBar.visibility = View.GONE
    }
```

Reply:
I sometimes do that too but I think it decreases the readability of the code.

Reply:
🤷‍♂️